horovod/common/operations.cc

```diff
@@ -541,53 +540,97 @@ void PerformOperation(TensorTable& tensor_table, Response response) {
+    stall_shutdown_time = std::chrono::seconds(0);
+  }
+
   for (auto& m : *state.message_table) {
     auto tensor_name = m.first;
     std::vector<Request>& messages = std::get<0>(m.second);
     std::chrono::steady_clock::time_point start_at = std::get<1>(m.second);
+    auto lag = now - start_at;
 
-    if (now - start_at > STALL_WARNING_TIME) {
-      std::stringstream message;
-      if (!preamble) {
-       message << "One or more tensors were submitted to be "
-                  "reduced, gathered or broadcasted by subset of ranks and "
-                  "are waiting for remainder of ranks for more than "
-               << std::chrono::duration_cast<std::chrono::seconds>(
-                   STALL_WARNING_TIME)
-                   .count()
-               << " seconds. "
-               << "This may indicate that different ranks are trying to "
-                  "submit different tensors or that only subset of ranks is "
-                  "submitting tensors, which will cause deadlock. "
-               << std::endl << "Stalled ops:" ;
-        preamble = true;
-      }
-      message << tensor_name;
-      message << " [missing ranks:";
+    if (lag > stall_warning_time) {
       std::unordered_set<int32_t> ready_ranks;
-      bool missing_preamble = false;
       for (auto msg_iter = messages.begin(); msg_iter != messages.end();
            ++msg_iter) {
         ready_ranks.insert(msg_iter->request_rank());
       }
+
       for (int32_t rank = 0; rank < state.size; ++rank) {
         if (ready_ranks.find(rank) == ready_ranks.end()) {
-          if (!missing_preamble) {
-            message << " ";
-            missing_preamble = true;
-          } else {
-            message << ", ";
+          missing_ranks[rank].insert(tensor_name);
+          if (stall_shutdown_time > std::chrono::seconds(0) && lag > stall_shutdown_time) {
+            shutdown_ranks.insert(rank);
+            should_shut_down = true;
           }
-          message << rank;
         }
       }
+    }
+  }
+
+  if (!missing_ranks.empty()) {
+    std::stringstream message;
+    message << "One or more tensors were submitted to be "
+               "reduced, gathered or broadcasted by subset of ranks and "
+               "are waiting for remainder of ranks for more than "
+            << stall_warning_time.count() << " seconds. "
+            << "This may indicate that different ranks are trying to "
+               "submit different tensors or that only subset of ranks is "
+               "submitting tensors, which will cause deadlock. "
+            << std::endl << "Stalled ranks:";
+    for (auto& kv: missing_ranks) {
+      message << std::endl << kv.first;
+      if (shutdown_ranks.find(kv.first) != shutdown_ranks.end()) {
+        message << "!";
+      }
+
+      message << ": [";
+      auto it = kv.second.begin();
+      message << *it;
+      while (++it != kv.second.end()) {
```

Comment:
One of the complaints from the user was that sometimes these messages are too long - we could probably just "..." after the first 5 tensors?

Reply:
done